smartsim/_core/_cli/plugin.py

```diff
@@ -0,0 +1,37 @@
+import argparse
+import shutil
+import subprocess as sp
+import typing as t
+
+from smartsim._core._cli.utils import MenuItemConfig
+
+
+plugins: t.List[t.Callable[[], MenuItemConfig]] = []
+
+
+def dynamic_execute(cmd: str) -> t.Callable[[argparse.Namespace, t.List[str]], int]:
+    def process_execute(_args: argparse.Namespace, unparsed_args: t.List[str]) -> int:
+        if not shutil.which(cmd):
+            raise ValueError(f"{cmd} plugin not found. Please ensure it is installed")
+        combined_cmd = [cmd] + unparsed_args
+        with sp.Popen(combined_cmd, stdout=sp.PIPE, stderr=sp.PIPE) as process:
+            stdout, _ = process.communicate()
+            while process.returncode is None:
+                stdout, _ = process.communicate()
+                print(stdout.decode("utf-8"))
+
+            return process.returncode
+
+    return process_execute
+
+
+def dashboard() -> MenuItemConfig:
+    return MenuItemConfig(
```

Comment:
I'm not sure that I like that first class commands are the same "type" as plugins... What do you think about this potential refactor:

```py
_TMenuItemHandler = Callable[[Namespace], int]
_TPluginItemHandler = Callable[[Namespace, t.List[str]], int]

@dataclass(frozen=True)
class MenuItemConfig:
    command: str
    description: str
    handler: _TMenuItemHandler
    configurator: t.Optional[_CliParserConfigurator] = None

    def execute(self, parser: argparse.ArgumentParser, args: Sequence[str]) -> None:
        ns = parser.parse_args(args)
        return self.handler(ns)


# TODO: I've not fully decided if these two classes should have
#     some common base class or protocol or something for
#     typing/static analysis purposes. I'll leave that to PR
#     author's discretion


@dataclass(frozen=True)
class MenuPluginConfig:
    command: str
    description: str
    handler: _TPluginItemHandler
    configurator: t.Optional[_CliParserConfigurator] = None  # <-- TODO: not strictly sure
                                                             #           if this is needed

    def execute(self, parser: argparse.ArgumentParser, args: Sequence[str]) -> None:
        ns, unparsed = parser.parse_known_args(args)
        # TODO: `ns` will always be empty if we do not allow for a
        #             `configurator`. Feel free to omit this line and just
        #             pass `args` directly if we do not need it!
        return self.handler(ns, unparsed)
```
And then in `SmartCli` we can remove the `SmartCli().args` attr (from what I can tell its not used outside of the `Smart().execute()` method anyway) and simply remove the `if menu_item.is_plugin` check and just directly call
```py
menu_item.execute(self.parser, app_args)
```
?

Pros: I _really_ like that first class items and plugins can be differentiated by type
Cons: Adds a new class/makes the whole CLI construction a hair more difficult :(

Reply:
Per offline conversation, we have decided to kick this into a new ticket

Reply:
@MattToast ^^ Make sure to do this since you made the suggestion!

Reply:
Created ticket 568